### ISSUE TITLE
Fine-tuning without freezing transformer parameter leads to poor performance

### ISSUE THREAD

**Issue by Zehui127** [lucidrains/enformer-pytorch#16]
Dear team,

As far as I can understand, the current fine-tuning feature will fix transformer model and only fine-tune the **linear heads**. I wonder has anyone try to fine-tuning the whole model without fixing the transformer parameter? I tried, but the performance is not good; for example, when I simply fine-tune the model on the original basenji training set, the correlation score decreased from 0.6 to 0.1. I wonder has anyone encounter the similar issue and what is the potential reason for this?

**Comment by lucidrains** [comment]
@Zehui127 it is difficult because of the batchnorms

however, recently it seems like researchers are giving good testimonies about using [LoRA](https://arxiv.org/abs/2106.09685)

i could add this as an option, on the condition that you let me know of the results once you try it?

**Comment by Zehui127** [comment]
@lucidrains Thanks for responding to this. It will be much appreciated if there is an option for doing tune-fining without freezing the transformer parameter. I will definitely let you know the result if you make the change.

**Comment by lucidrains** [comment]
@lucidrains oh, it has been a while and i have forgotten what i built here..

have you tried [this option](https://github.com/lucidrains/enformer-pytorch/blob/main/enformer_pytorch/finetune.py#L127) yet

i'm not really caught up on the fine tuning literature, but my impression is that you don't really need to unfreeze the entire network

**Comment by lucidrains** [comment]
@Zehui127 oh, it has been a while and i have forgotten what i had built

have you tried [this option](https://github.com/lucidrains/enformer-pytorch/blob/main/enformer_pytorch/finetune.py#L127) yet

i'm not really caught up on the fine tuning literature, but my impression is that you don't really need to unfreeze the entire network

**Comment by lucidrains** [comment]
@Zehui127 LoRA should be quite good, i see people doing dreambooth fine tuning in stable diffusion using that technique

**Comment by lucidrains** [comment]
that option unfreezes the layernorms of the transformer, a popular technique. there's also another option to unfreeze the penultimate layers of the transformer

**Comment by lucidrains** [comment]
oh actually nevermind, I never integrated that last N layers fine-tuning thing

what is your batch size when fine-tuning?

**Comment by lucidrains** [comment]
yeah, I think unfreezing the transformer may be your best bet, since it is free of batch normalization (while keeping conv stem frozen)

**Comment by lucidrains** [comment]
I'll take a look tomorrow morning and see if I can easily squeeze in the unfreezing of last N layers of transformer setting, so you can play around with that

**Comment by lucidrains** [comment]
may I ask whose lab you are working for?

**Comment by Zehui127** [comment]
So which means that the following will work right?
1.  if I train the model with batch size more than 32, and unfroze the whole network (conv + transformer)
2. If I train the model with batch size less than 32, and froze the conv + unfroze the transformer

**Comment by Zehui127** [comment]
I'm a PhD in Imperial College London

**Comment by lucidrains** [comment]
well, you forget that catastrophic forgetting can still happen, esp if you unfreeze the whole network at once. recommend you google for literature on that and see what the latest bag if tricks are

higher batch sizes will help with the batch norm yeah

**Comment by Zehui127** [comment]
Thanks for your kind help!

**Comment by lucidrains** [comment]
ahh ok cool, yeah then by your tomorrow afternoon, I can probably get the unfreezing of entire or last N layers of transformers finished

**Comment by lucidrains** [comment]
no problem, I want this repository to be useful 🌝

**Comment by lucidrains** [comment]
@Zehui127 Hi Zehui

reviewed the code this morning and noticed that i already do freeze the batchnorms, so if you are using any one of the finetuning adapters, the batchnorm should not be the issue (as long as you are gradient accumulating properly and not training on batch size of 2). in the case you did things correctly, likely issue would be catastrophic forgetting

i've finished integrating the feature for freezing the entire network but the last N layers of the transformer. you can use it by setting [this keyword argument](https://github.com/lucidrains/enformer-pytorch/commit/a8d8d2fba6fd59f0f3755bef6fbf4bf1dfb0097b#diff-8b377aa9e5b0fd2c166fb57e3cfc1e608f2593dc29561bd4c262cbaea410f4c3R183) on adapter forward to an integer

**Comment by lucidrains** [comment]
@Zehui127 which adapter are you using for the finetuning? the head adapter?

**Comment by Zehui127** [comment]
@lucidrains Thanks for helping on this. I was using the HeadAdapterWrapper before.

**Comment by lucidrains** [comment]
@Zehui127 were you gradient accumulating when training on batch size of 2? wondering if i should offer a training wrapper that takes care of this too (noticed i actually left myself a todo for researchers like you, but never completed it)

**Comment by lucidrains** [comment]
@Zehui127 also, have you tried the discrete key / value bottleneck feature? https://arxiv.org/abs/2207.11240 i quite liked the paper, but never tried it out on a task yet

**Comment by Zehui127** [comment]
@lucidrains ， It looks like a very interesting paper, I will try and let you know the result. For  gradient accumulation, previously, I did use it in some runs, but it didn't have much improvement. I'm refactoring the code at this moment to allow larger effective batch size. Hopefully, with all these changes, the performance could be improved. Thanks again.

**Comment by lucidrains** [comment]
@Zehui127 ok, keep me in the know!

**Comment by lucidrains** [comment]
@Zehui127 you'll need gradient accumulation for large effective batch size
